feat: Add `reportUsedIgnorePattern` option to `no-unused-vars` rule

Pearce-Ropion opened the pull request:
<!--
    Thank you for contributing!

    ESLint adheres to the [JS Foundation Code of Conduct](https://eslint.org/conduct).
-->

#### Prerequisites checklist

- [x] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/HEAD/CONTRIBUTING.md).

#### What is the purpose of this pull request? (put an "X" next to an item)

<!--
    The following template is intentionally not a markdown checkbox list for the reasons
    explained in https://github.com/eslint/eslint/pull/12848#issuecomment-580302888
-->

- [x] Documentation update
- [ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/bug-report.md))
- [ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-proposal.md))
- [x] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-change-proposal.md))
- [ ] Add autofix to a rule
- [ ] Add a CLI option
- [ ] Add something to the core
- [ ] Other, please explain:

<!--
    If the item you've checked above has a template, please paste the template questions below and answer them. (If this pull request is addressing an issue, you can just paste a link to the issue here instead.)
-->

**What rule do you want to change?**

`no-unused-vars`

**What change do you want to make (place an "X" next to just one item)?**

- [x] Generate more warnings
- [ ] Generate fewer warnings
- [ ] Implement autofix
- [ ] Implement suggestions

**How will the change be implemented (place an "X" next to just one item)?**

- [x] A new option
- [ ] A new default behavior
- [ ] Other

**Please provide some example code that this change will affect:**

Example 1:
```js
/* eslint no-unused-vars: ["error", { "varsIgnorePattern": "^_" }] */

const _a = 5;
const _b = _a + 5
```

Example 2:
```js
/* eslint no-unused-vars: ["error", { "argsIgnorePattern": "^_" }] */

function foo(_a) {
  return _a + 5
}
```

Example 3:
```js
/* eslint no-unused-vars: ["error", { "destructuredArrayIgnorePattern": "^_" }] */

const [ a, _b ] = items;
console.log(a + _b);
```

Example 4:
```js
/* eslint no-unused-vars: ["error", { "caughtErrorsIgnorePattern": "^_" }] */

try {}
catch(_err) {
  console.error(_err);
}
```

**What does the rule currently do for this code?**

In all examples, no warnings are reported.

**What will the rule do after it's changed?**

When the new option `reportUsedIgnorePattern` is enabled, each of the examples above will report; indicating that a variable that has been marked as an ignored unused variable is being used.

<!--
    Please ensure your pull request is ready:

    - Read the pull request guide (https://eslint.org/docs/latest/contribute/pull-requests)
    - Include tests for this change
    - Update documentation for this change (if appropriate)
-->

<!--
    The following is required for all pull requests:
-->

#### What changes did you make? (Give an overview)

This change adds a new option `reportUsedIgnorePattern` to the `no-unused-vars` rule. When enabled, the rule will report any instances of variables that match any of the valid ignore pattern options when that variable is actually being used. 

Resolves: #17568

#### Is there anything you'd like reviewers to focus on?

When finding all references in order to report on the correct nodes, it is also finding the variable declaration. So a question for the team would be which nodes should this option report on:
* Should all references including the declaration be reported?
* Should only the variable declaration be reported? 
* Should only references not including the declaration be reported

<!-- markdownlint-disable-file MD004 -->

costinsin commented on lib/rules/no-unused-vars.js:
```diff
@@ -178,6 +191,47 @@ module.exports = {
+                let type;
+                let pattern;
+
+                if (defType === "CatchClause" && config.caughtErrorsIgnorePattern) {
+                    type = "args";
+                    pattern = config.caughtErrorsIgnorePattern.toString();
+                } else if (defType === "Parameter" && config.argsIgnorePattern) {
+                    type = "args";
+                    pattern = config.argsIgnorePattern.toString();
+                } else if (defType !== "Parameter" && config.varsIgnorePattern) {
+                    type = "vars";
+                    pattern = config.varsIgnorePattern.toString();
+                } else if (def.name.parent.type === "ArrayPattern" && config.destructuredArrayIgnorePattern) {
+                    type = "elements of array destructuring";
+                    pattern = config.destructuredArrayIgnorePattern.toString();
+                }
```
```suggestion
                switch (true) {
                    case defType === "CatchClause" && config.caughtErrorsIgnorePattern:
                        type = "args";
                        pattern = config.caughtErrorsIgnorePattern.toString();
                        break;
                    case defType === "Parameter" && config.argsIgnorePattern:
                        type = "args";
                        pattern = config.argsIgnorePattern.toString();
                        break;
                    case defType !== "Parameter" && config.varsIgnorePattern:
                        type = "vars";
                        pattern = config.varsIgnorePattern.toString();
                    case def.name.parent.type === "ArrayPattern" && config.destructuredArrayIgnorePattern:
                        type = "elements of array destructuring";
                        pattern = config.destructuredArrayIgnorePattern.toString();
                        break;
                }
```

Pearce-Ropion replied:
This seems like an anti pattern.

Also, there are 2 other functions that are very similar to this one in this rule already so making this change would be inconsistent with how the rest of the rule is formatted.

costinsin replied:
This is clearly not an anti pattern, as TypeScript 5.3 [just added](https://devblogs.microsoft.com/typescript/announcing-typescript-5-3-beta/#switch-true-narrowing#switch-true-narrowing) type narrowing for this pattern. I feel it resembles the powerful `match` present usually in functional languages.

This being said, I agree with you that we should keep the same code style and not mix them. It was just a suggestion.

mdjermanovic commented:
> * Should all references including the declaration be reported?
> * Should only the variable declaration be reported?
> * Should only references not including the declaration be reported

Would it make sense to report only on the reasons the variable is considered used, which are usually its read references?

This is how it currently works:

```js
/* eslint no-unused-vars: ["error", {
    varsIgnorePattern: "^_",
    reportUsedIgnorePattern: true
}] */

var _a; // not reported

_a = 1; // this is reported but doesn't seem relevant?

foo(_a); // this is also reported, ok

```

But this variable is considered used only because of the read reference so it might make sense to report like this:

```js
/* eslint no-unused-vars: ["error", {
    varsIgnorePattern: "^_",
    reportUsedIgnorePattern: true
}] */

var _a; // not reported

_a = 1; // not reported

foo(_a); // reported

```

mdjermanovic commented on lib/rules/no-unused-vars.js:
```diff
@@ -178,6 +191,47 @@ module.exports = {
+                }
+
+                if (type) {
+                    additional = `. Allowed unused ${type} must not match ${pattern}`;
```
```suggestion
                    additional = `. Used ${type} must not match ${pattern}`;
```

Pearce-Ropion commented:
> > * Should all references including the declaration be reported?
> > * Should only the variable declaration be reported?
> > * Should only references not including the declaration be reported
> 
> Would it make sense to report only on the reasons the variable is considered used, which are usually its read references?
> 
> This is how it currently works:
> 
> ```js
> /* eslint no-unused-vars: ["error", {
>     varsIgnorePattern: "^_",
>     reportUsedIgnorePattern: true
> }] */
> 
> var _a; // not reported
> 
> _a = 1; // this is reported but doesn't seem relevant?
> 
> foo(_a); // this is also reported, ok
> ```
> 
> But this variable is considered used only because of the read reference so it might make sense to report like this:
> 
> ```js
> /* eslint no-unused-vars: ["error", {
>     varsIgnorePattern: "^_",
>     reportUsedIgnorePattern: true
> }] */
> 
> var _a; // not reported
> 
> _a = 1; // not reported
> 
> foo(_a); // reported
> ```

That makes sense to me. However, I would also expect that a user shouldn't be allowed to write to an ignored unused variable since it should be unused.

mdjermanovic commented on lib/rules/no-unused-vars.js:
```diff
@@ -711,8 +789,25 @@ module.exports = {
+                    for (const usedIgnoredVar of usedIgnoredVars) {
+                        if (usedIgnoredVar.defs.length > 0) {
+                            const references = usedIgnoredVar.references.filter(
+                                ref => ref.isRead() && ref.from.variableScope === usedIgnoredVar.scope.variableScope
```
Per this condition, `_a` wouldn't be reported at all in the following example, which seems wrong as it's used?

```js
/* eslint no-unused-vars: ["error", {
    varsIgnorePattern: "^_",
    reportUsedIgnorePattern: true
}]*/

var _a = 42;

foo(() => _a);

```

Pearce-Ropion replied:
Do you think I should revert to just `ref => ref.from.variableScope === usedIgnoredVar.scope.variableScope`? Is there a better condition to use?

Pearce-Ropion commented:
I changed it to skip all variable initializations but otherwise report on all references. I also doesn't check whether the variable scope matches anymore so it can handle variables defined in the outer scope which are used in an inner scope.

Pearce-Ropion commented:
@mdjermanovic I think this is working as expected now. Can you think of any other cases that I could test for? Otherwise, I think we are good to go.

mdjermanovic commented:
@Pearce-Ropion apologies for the delay in reviewing this PR.

> I changed it to skip all variable initializations but otherwise report on all references.

I'm not sure what the common use cases for the `varsIgnorePattern` option are, but if user has a code like the following where the rule doesn't report any errors:

```js
/*eslint no-unused-vars: ["error", { varsIgnorePattern: "^_", reportUsedIgnorePattern: true }]*/

var _foo;

_foo = 5;

```

and then they add more code that causes the variable to be considered used, so that the rule now reports it:

```js
/*eslint no-unused-vars: ["error", { varsIgnorePattern: "^_", reportUsedIgnorePattern: true }]*/

var _foo;

_foo = 5;

bar(_foo);

```

then, I think it makes the most sense to report only on the references where this rule considers the variable to be used: `_foo` in `bar(_foo)`, because `_foo = 5` has no effect on whether the variable is considered used, and it was valid before `bar(_foo)` was added.

Alternatively, we could report only the declaration assuming that the variable name is wrong because it was intended to be used (as opposed to reporting references, in which case the assumption is that the variable that wasn't intended to be used is mistakenly used).

Curious what others think about this.

Pearce-Ropion commented:
No worries, eslint is a massive project and I don't want to burden anyone. 

> then, I think it makes the most sense to report only on the references where this rule considers the variable to be used: `_foo` in `bar(_foo)`, because `_foo = 5` has no effect on whether the variable is considered used, and it was valid before `bar(_foo)` was added.
 
> Alternatively, we could report only the declaration assuming that the variable name is wrong because it was intended to be used (as opposed to reporting references, in which case the assumption is that the variable that wasn't intended to be used is mistakenly used).

Using vscode's eslint extension as an example, on one hand it would be useful to see reported issues on the usages since that's what is considered invalid according to the new option. However, It would be more clear what the problem is and how to fix it at the declaration, so it'd probably be better to do it there.

nzakas commented:
> then, I think it makes the most sense to report only on the references where this rule considers the variable to be used: _foo in bar(_foo), because _foo = 5 has no effect on whether the variable is considered used, and it was valid before bar(_foo) was added.

I think this makes the most sense. :+1:

@Pearce-Ropion sorry for the delay, we've been busy working on v9. Can you please rebase and make any further changes? Thanks!

nzakas commented:
Oops, accidentally hit "close"

Pearce-Ropion commented:
> Oops, accidentally hit "close"

lol, I was going to reopen once I pushed the changes.

Pearce-Ropion commented:
This is still active. Waiting for a review.

Pearce-Ropion commented on lib/rules/no-unused-vars.js:
```diff
@@ -178,6 +191,47 @@ module.exports = {
+                let type;
+                let pattern;
+
+                if (defType === "CatchClause" && config.caughtErrorsIgnorePattern) {
```
As per https://github.com/eslint/eslint/pull/17662#discussion_r1365318959 we opted to leave the if statements as is to reduce the scope of the PR and to maintain consistency with other similar functions within the rule. If I make this change, should I also update the other functions that use a similar flow?

Pearce-Ropion replied:
On second thought, given that this is the second time this suggestion has been made, I've gone ahead and refactored this and the other message getters in the rule for consistency to use `switch` statements.

nzakas replied:
Please double-check who the person is who is leaving suggestions on a PR. Not everyone is a team member. :)

nzakas commented on lib/rules/no-unused-vars.js:
```diff
@@ -144,22 +144,40 @@ module.exports = {
+                let pattern;
 
-            const additional = type ? `. Allowed unused ${type} must match ${pattern}` : "";
+                switch (def.type) {
+                    case "CatchClause":
+                        if (config.caughtErrorsIgnorePattern) {
+                            type = "args";
+                            pattern = config.caughtErrorsIgnorePattern;
+                        }
+                        break;
+
+                    case "Parameter":
+                        if (config.argsIgnorePattern) {
+                            type = "args";
+                            pattern = config.argsIgnorePattern;
+                        }
+                        break;
+
+                    default:
+                        if (config.varsIgnorePattern) {
+                            type = "vars";
+                            pattern = config.varsIgnorePattern;
+                        }
+                        break;
```
Nice :+1:

nzakas reviewed (approved):
LGTM. Would like @mdjermanovic to re-review before merging.

mdjermanovic commented:
In this case, the message seems misleading:

```js
/*eslint no-unused-vars: ["error", {
    "reportUsedIgnorePattern": true,
    "destructuredArrayIgnorePattern": "^_",
    "varsIgnorePattern": "[iI]gnored"
}]*/

let _x; // '_x' is marked as ignored but is used. Used vars must not match /[iI]gnored/u.

[_x] = arr;

foo(_x);

```

Tanujkanti4441 commented:
Hi @Pearce-Ropion, some changes are requested. you can check out.

Pearce-Ropion commented on tests/lib/rules/no-unused-vars.js:
```diff
@@ -442,6 +460,22 @@ ruleTester.run("no-unused-vars", rule, {
+        // https://github.com/eslint/eslint/issues/17568
+        {
+            code: "const a = 5; const _c = a + 5;",
+            options: [{ args: "all", varsIgnorePattern: "^_", reportUsedIgnorePattern: true }],
```
Yep, nice catch. I also apparently got the error output correct on the invalid cases.

mdjermanovic commented on lib/rules/no-unused-vars.js:
```diff
@@ -92,7 +103,8 @@ module.exports = {
             ignoreRestSiblings: false,
-            caughtErrors: "all"
+            caughtErrors: "all",
+            reportUsedIgnorePattern: false,
```
```suggestion
            reportUsedIgnorePattern: false
```

To fix the lint error.

mdjermanovic commented:
> In this case, the message seems misleading:
> 
> ```js
> /*eslint no-unused-vars: ["error", {
>     "reportUsedIgnorePattern": true,
>     "destructuredArrayIgnorePattern": "^_",
>     "varsIgnorePattern": "[iI]gnored"
> }]*/
> 
> let _x; // '_x' is marked as ignored but is used. Used vars must not match /[iI]gnored/u.
> 
> [_x] = arr;
> 
> foo(_x);
> ```

I'm still getting the same error, looks like this case still hasn't been addressed?

mdjermanovic commented:
Here's a similar case that should also be fixed:

```js
/*eslint no-unused-vars: ["error", {
    "reportUsedIgnorePattern": true,
    "destructuredArrayIgnorePattern": "^_",
    "varsIgnorePattern": "[iI]gnored"
}]*/

const [ignored] = arr; // 'ignored' is marked as ignored but is used. Used elements of array destructuring must not match /^_/u.

foo(ignored);

```

The message should be: `'ignored' is marked as ignored but is used. Used vars must not match /[iI]gnored/u.`

mdjermanovic commented on lib/rules/no-unused-vars.js:
```diff
@@ -610,6 +717,10 @@ module.exports = {
                             config.destructuredArrayIgnorePattern &&
                             config.destructuredArrayIgnorePattern.test(def.name.name)
                         ) {
+                            if (config.reportUsedIgnorePattern && isUsedVariable(variable)) {
+                                usedIgnoredVars.push(variable);
+                            }
```
Instead of collecting these vars and later figuring out why they were considered as used ignored, we could just report them right away with the data we have at this point:

```suggestion
                            if (config.reportUsedIgnorePattern && isUsedVariable(variable)) {
                                context.report({
                                    node: def.name,
                                    messageId: "usedIgnoredVar",
                                    data: {
                                        varName: variable.name,
                                        additional: `. Used elements of array destructuring must not match ${config.destructuredArrayIgnorePattern}`
                                    }
                                });
                            }
```

(the same for catch arguments, params and variables).

That would fix https://github.com/eslint/eslint/pull/17662#issuecomment-1910344522 and https://github.com/eslint/eslint/pull/17662#issuecomment-1959062359.

nzakas commented:
@Pearce-Ropion are you still working on this?

Pearce-Ropion commented:
> @Pearce-Ropion are you still working on this?

Yes, apologies. I've been slammed with work. I'll get the proposed suggestions in this weekend

Pearce-Ropion commented:
Ok, I've applied @mdjermanovic's suggestions to report as soon as we detect the used ignored variable. I also explicitly added the test cases that @mdjermanovic used.

I've also updated the message data handling. I had already updated the other message data handlers as per previous comments, but I noticed that all 3 of the message data getters were doing basically the same thing. So I've added a `getVariableDescription` helper function to get the required info for the message data base on the type of the variable that is currently being reported.

nzakas reviewed (approved):
LGTM. Would like @mdjermanovic to review before merging.

mdjermanovic reviewed (approved):
LGTM, thanks for contributing!
